### PR TITLE
[#IOCIT-118] make prevalidate errors more readable

### PR DESCRIPTION
**Pull request by arcogabbo**
#### List of Changes

* refactored the prevalidate method to generate more readable errors, including `requestId` and `idpIssuer` fields

#### Motivation and Context

when debugging a login issue, it can be useful to have more data to better understand the problem

#### How Has This Been Tested?

build and test passed

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/utils/saml.ts [hunk 1]
```diff
@@ -68,7 +121,7 @@ export const getPreValidateResponse = (
   doneCb,
   callback
   // eslint-disable-next-line sonarjs/cognitive-complexity
-): Promise<void | E.Left<void> | E.Right<void>> => {
+): Promise<void | E.Either<void, void>> => {
```
Comment (balanza) [inline src/utils/saml.ts:124]: It's out of scope for this PR, but `Promise<Either<L,R>>` isn't `TaskEither<L,R>`?

Also, what `Promise<void>` stands for?

File: src/utils/saml.ts [hunk 2]
```diff
@@ -81,72 +134,70 @@ export const getPreValidateResponse = (
     "Response"
   );
 
+  const maybeIdpIssuer = getSamlIssuer(doc);
+
   const hasStrictValidation = pipe(
     O.fromNullable(strictValidationOptions),
     O.chain(_ =>
```
Comment (balanza) [inline src/utils/saml.ts:141]: Please give `_` a meaningful name, or @michaeldisaro gets mad

**Review by balanza — COMMENTED** [review]
Implementation looks fine to me. I'd be more explicit about:
1. the kind of data we are logging to help privacy team to audit our code 
2. the queries we expect to do on logs once we have such data -  and the ones we **don't** want to be able to do

Both can be achieved by commenting out the code imho

File: src/utils/saml.ts (hunk 1 above)
Comment (arcogabbo) [reply src/utils/saml.ts:124]: you are right, it is a `TaskEither<void,void>` in this case. should I do this change in another PR? it was a small catch I saw while refactoring

File: src/utils/saml.ts (hunk 2 above)
Comment (arcogabbo) [reply src/utils/saml.ts:141]: what do you think about `validationOptions` for that parameter?

File: src/utils/saml.ts (hunk 1 above)
Comment (BurnedMarshal) [reply src/utils/saml.ts:124]: There is a small difference from `Promise<Either<L,R>>` and `TaskEither<L,R>`, the task either operation *can't fail* instead the Promise can do. The conversion can be made catching the promise error and putting it on the left `TaskEither<E|L, R>`.

Comment (balanza) [reply src/utils/saml.ts:124]: You are right, indeed `TaskEither<L,R> = () => Promise<Either<L,R>>`

Dumb request I made, sorry

**Review by balanza — APPROVED** [review]

File: src/utils/saml.ts [hunk 3]
```diff
@@ -81,72 +134,70 @@ export const getPreValidateResponse = (
...
+    { readonly InResponseTo: NonEmptyString; readonly Response: Element }
+  > = pipe(
+    responsesCollection,
+    E.fromPredicate(
+      coll => coll.length < 2,
+      _ => new Error("SAML Response must have only one Response element")
+    ),
+    E.map(coll => coll.item(0)),
+    E.chain(
+      E.fromNullable(new Error("Missing Response element inside SAML Response"))
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/utils/saml.ts:166]: based on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection) the item method returns `null` if the index argument is out of bound(makin g the return type `Element | null`). this chain is removing the `null` from the variable type. I agree that this check can be done before when checking the length

File: src/utils/saml.ts [hunk 4]
```diff
@@ -81,72 +134,70 @@ export const getPreValidateResponse = (
...
-    readonly RequestAuthnRequest: Element;
-  };
+  const errorOrResponse: E.Either<
+    Error,
+    { readonly InResponseTo: NonEmptyString; readonly Response: Element }
+  > = pipe(
+    responsesCollection,
+    E.fromPredicate(
+      coll => coll.length < 2,
+      _ => new Error("SAML Response must have only one Response element")
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/utils/saml.ts:162]: The next step check if the response is missing with the error `Missing Reponse element inside SAML Response`. If this check will be `coll.length === 1` the next check never be executed.

File: src/utils/saml.ts [hunk 5]
```diff
@@ -81,72 +134,70 @@ export const getPreValidateResponse = (
...
+    ),
+    E.map(coll => coll.item(0)),
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/utils/saml.ts:164]: See [here](https://github.com/pagopa/io-spid-commons/pull/119#discussion_r980836328)

File: src/utils/saml.ts [hunk 6]
```diff
@@ -81,72 +134,70 @@ export const getPreValidateResponse = (
...
+    E.chain(
+      E.fromNullable(new Error("Missing Response element inside SAML Response"))
+    ),
+    E.chain(Response =>
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/utils/saml.ts:171]: `PascalCase` is enabled for variable naming convention, more details [here](https://github.com/pagopa/eslint-rules/blob/b157c7d0ccc88fb2f0c471fd69dc4234aa17d356/strong.js#L48)

File: src/utils/saml.ts [hunk 7]
```diff
@@ -68,7 +121,7 @@ export const getPreValidateResponse = (
   doneCb,
   callback
   // eslint-disable-next-line sonarjs/cognitive-complexity
-): Promise<void | E.Left<void> | E.Right<void>> => {
+): void => {
```
Comment (arcogabbo) [inline src/utils/saml.ts:124]: the type `PreValidateResponseT` explicitly says that this function is returning void, indicating that the return edvalue is not used. I procedeed and edited this as a `void` returning function, as the type suggests.

**Review by michaeldisaro — APPROVED** [review]
LGTM

File: src/utils/saml.ts [hunk 8]
```diff
@@ -81,72 +134,73 @@ export const getPreValidateResponse = (
...
+    )
+  );
 
-  type IIssueInstantWithAuthnContextCR = IIssueInstant & {
-    readonly RequestAuthnContextClassRef: NonEmptyString;
-  };
+  const requestId: string = pipe(
+    errorOrPartiallyValidatedResponse,
+    E.map(({ InResponseTo }) => InResponseTo),
+    E.getOrElse(() => "not available")
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:185]: ditto

File: src/utils/saml.ts [hunk 9]
```diff
@@ -564,26 +609,37 @@ export const getPreValidateResponse = (
...
+          name: "spid.error.transformOccurenceOverflow",
+          type: "ERROR"
+        });
+      } else {
+        const samlError: ISAMLError = toSamlError(error.message);
+        eventHandler({
+          data: {
+            idpIssuer: samlError.idpIssuer,
+            message: samlError.message,
+            requestId: samlError.requestId
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:637]: the `toSamlError` function is used only here. What is his purpose?
Will be the same this version
```suggestion
        const samlError: ISAMLError = toSamlError(error.message);
        eventHandler({
          data: {
            idpIssuer,
            message: error.message,
            requestId
```

**Review by BurnedMarshal — COMMENTED** [review]
This suggestion can be applied if needed to the next PR

File: src/utils/saml.ts (hunk 9 above)
Comment (arcogabbo) [reply src/utils/saml.ts:637]: fixed in https://github.com/pagopa/io-spid-commons/pull/121/commits/8a663a2b24774f521a9a08720d21d5a348d7a998

File: src/utils/saml.ts (hunk 8 above)
Comment (arcogabbo) [reply src/utils/saml.ts:185]: fixed in https://github.com/pagopa/io-spid-commons/pull/121/commits/8a663a2b24774f521a9a08720d21d5a348d7a998

File: src/utils/saml.ts [hunk 10]
```diff
@@ -81,72 +134,73 @@ export const getPreValidateResponse = (
...
-  }
-
-  type IRequestAndResponseStep = IBaseOutput & {
-    readonly SAMLRequestCache: ISamlCacheType;
-  };
-
-  type ISAMLRequest = IRequestAndResponseStep & { readonly Request: Document };
+  const idpIssuer: string = pipe(
+    maybeIdpIssuer,
+    O.getOrElse(() => "not available")
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/utils/saml.ts:152]: fixed in https://github.com/pagopa/io-spid-commons/pull/121/commits/8a663a2b24774f521a9a08720d21d5a348d7a998

File: src/utils/saml.ts [hunk 11]
```diff
@@ -56,6 +56,59 @@ export {
...
+  readonly numberOfTransforms: number;
+}
+
+export const errorToSamlError = (
+  e: Error,
+  info?: { readonly requestId?: string; readonly idpIssuer?: string }
+): ISAMLError => ({
+  // the nullish coalescing operator and optional chaining will cover all cases here
+  // eslint-disable-next-line sonarjs/no-duplicate-string
+  idpIssuer: info?.idpIssuer ?? "not available",
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/utils/saml.ts:105]: fixed in https://github.com/pagopa/io-spid-commons/pull/121/commits/8a663a2b24774f521a9a08720d21d5a348d7a998

**Pull request merged**